src/downloader.ts

```diff
@@ -9,57 +10,62 @@ import semver from 'semver'
 import { Logger } from './logger.js'
 import { getBinaryExtension, getPlatformTarget } from './platform.js'
 
-const download = async (targetDirectory: string, versionRange: string, logger: Logger) => {
+const downloadWithRetry = async (targetDirectory: string, versionRange: string, logger: Logger) =>
+  await pRetry(async () => await download(targetDirectory, versionRange), {
+    retries: 3,
+    onFailedAttempt: (error) => {
+      logger.system('Deno download with retry failed', error)
+    },
+  })
+
+const download = async (targetDirectory: string, versionRange: string) => {
   const zipPath = path.join(targetDirectory, 'deno-cli-latest.zip')
-  const data = await downloadVersionWithRetry(versionRange, logger)
+  const data = await downloadVersion(versionRange)
   const binaryName = `deno${getBinaryExtension()}`
   const binaryPath = path.join(targetDirectory, binaryName)
-  const file = fs.createWriteStream(zipPath)
-
-  await new Promise((resolve, reject) => {
-    data.pipe(file)
-    data.on('error', reject)
-    file.on('finish', resolve)
-  })
-
-  await extractBinaryFromZip(zipPath, binaryPath, binaryName)
+  const file = createWriteStream(zipPath)
 
   try {
-    await fs.promises.unlink(zipPath)
-  } catch {
-    // no-op
+    await new Promise((resolve, reject) => {
+      data.pipe(file)
+      data.on('error', reject)
+      file.on('finish', resolve)
+    })
+
+    await extractBinaryFromZip(zipPath, binaryPath, binaryName)
+
+    return binaryPath
+  } finally {
+    // Try closing and deleting the zip file in an case, error or not
+    await promisify(file.close.bind(file))()
```

Comment:
Is there a chance that this not fire and we'll await indefinitely?

Reply:
Well I hope not. And that would be very unexpected behavior. If an error happens, then the `callback(err)` would be called -> reject.